docs/spec.rst

```diff
@@ -0,0 +1,498 @@
+   regardless of its bound):
+   [`#17 <https://github.com/jspecify/jspecify/issues/17>`__]
+
+   -  Return type of a method.
+   -  Formal parameter type (of a method, constructor, or lambda
+      expression).
+   -  Field type.
+   -  Type parameter upper bound (but *not* the type parameter itself)
+      [`#60 <https://github.com/jspecify/jspecify/issues/60>`__].
+
+-  Illegal when applied to a class declaration.
+   [`#7 <https://github.com/jspecify/jspecify/issues/7>`__]
+
+   -  For example, ``public @Nullable class Foo {}`` is not allowed.
+
+-  Except as already indicated, the legality of applying to a type usage
+   in `implementation
+   code <https://docs.google.com/document/d/1KQrBxwaVIPIac_6SCf--w-vZBeHkTvtaqPSU_icIccc/edit#bookmark=id.cjuxrgo7keqs>`__
+   (and the semantics of doing so) is currently unspecified:
+
+   -  A local variable type.
+   -  The type in a cast expression.
+   -  An array or object creation expression.
+   -  An explicit type argument supplied to a generic method or
+      constructor (including via a member reference), or to an instance
+      creation expression for a generic class.
+
+-  Applying more than one of these annotations to the same `type
+   component <https://docs.google.com/document/d/1KQrBxwaVIPIac_6SCf--w-vZBeHkTvtaqPSU_icIccc/edit#bookmark=kix.g7gl9fwq1tt5>`__
+   within the same type usage (since we currently define only
+   ``@Nullable``, this means applying ``@Nullable`` twice) is illegal.
+   [*debated*
+   `#100 <https://github.com/jspecify/jspecify/issues/100>`__]
+
+Tools are encouraged to treat an illegally placed annotation in Java
+source code as an error. In bytecode, illegally placed annotations may
+be best ignored.
+
+Type usages defined here as *legal* or *illegal* (as opposed to
+*unspecified*) to receive type use annotations are considered **in
+scope** of this specification. Tools may extend this specification to
+*unspecified* type usages at their own choosing.
+
+The defaulting annotations
+--------------------------
+
+In this release, we will provide a single parameterless declaration
+annotation called ``@DefaultNonNull``.
+[`#5 <https://github.com/jspecify/jspecify/issues/5>`__,
+`#87 <https://github.com/jspecify/jspecify/issues/87>`__]
+
+.. _structural-legality-defaulting:
+
+Structural legality
+~~~~~~~~~~~~~~~~~~~
+
+-  The set of defaulting annotations can be applied to:
+
+   -  A *named* class.
+   -  A package. [*debated*
+      `#34 <https://github.com/jspecify/jspecify/issues/34>`__]
+   -  A module. [*debated*
+      `#34 <https://github.com/jspecify/jspecify/issues/34>`__]
+   -  *Not* a method
+      [`#43 <https://github.com/jspecify/jspecify/issues/43>`__],
+      constructor
+      [`#43 <https://github.com/jspecify/jspecify/issues/43>`__], or
+      field [`#50 <https://github.com/jspecify/jspecify/issues/50>`__].
+      *debated*
+
+-  Applying more than one of these annotations to the same scope (since
```

Comment:
nit: maybe we can lose this rule (for both @Nullable and @DefaultNonNull), since javac will reject duplicate use anyway.

Reply:
I think I made the same comment on the Docs version of this at one point :) As I recall, kevinb took the position that this part of the spec was essentially saying "These annotations will not be `@Repeated`." While it's hard to imagine that anyone would even think to define them that way, I suppose it's technically something that we want to forbid.